dali/pipeline/pipeline.cc

```diff
@@ -711,25 +711,101 @@ OpNode * Pipeline::GetOperatorNode(const std::string& name) {
   return &(graph_.Node(name));
 }
 
-std::map<std::string, Index> Pipeline::EpochSize() {
+std::map<std::string, Index> Pipeline::EpochSize(bool consider_padding) {
   std::map<std::string, Index> ret;
   for (Index i = 0; i < graph_.NumOp(OpType::CPU); ++i) {
     const OpNode &current = graph_.Node(OpType::CPU, i);
-    Index epoch_size = current.op->epoch_size();
+    Index epoch_size = current.op->epoch_size(consider_padding);
     if (epoch_size != -1) {
       ret.insert(make_pair(current.instance_name, epoch_size));
     }
   }
   for (Index i = 0; i < graph_.NumOp(OpType::GPU); ++i) {
     const OpNode &current = graph_.Node(OpType::GPU, i);
-    Index epoch_size = current.op->epoch_size();
+    Index epoch_size = current.op->epoch_size(consider_padding);
     if (epoch_size != -1) {
       ret.insert(make_pair(current.instance_name, epoch_size));
     }
   }
   return ret;
 }
 
+std::map<std::string, int> Pipeline::NumberOfShards() {
+  std::map<std::string, int> ret;
+  for (Index i = 0; i < graph_.NumOp(OpType::CPU); ++i) {
+    const OpNode &current = graph_.Node(OpType::CPU, i);
+    int number_of_shards = current.op->number_of_shards();
+    if (number_of_shards != -1) {
+      ret.insert(make_pair(current.instance_name, number_of_shards));
+    }
+  }
+  for (Index i = 0; i < graph_.NumOp(OpType::GPU); ++i) {
+    const OpNode &current = graph_.Node(OpType::GPU, i);
+    int number_of_shards = current.op->number_of_shards();
+    if (number_of_shards != -1) {
+      ret.insert(make_pair(current.instance_name, number_of_shards));
+    }
+  }
+  return ret;
+}
```

Comment:
There is `OpNode& Node(OpNodeId id)` that can lookup any node in the graph regardless of its stage, so you could have handled this with one loop. Still, the amount of boilerplate suggests, that there should be better way of handling this.

Either pack it all together as I suggested earlier, or maybe use a query mechanism like in #1880 (I know this was added more recently).

Reply:
I don't think that #1880 way of extracting operators internals is the way that operator want to follow to expose arbitrary data (we may want to modify in the getter).

Reply:
Also, do we really want to build a map every time we access some info about state of one Operator? Maybe we should just allow to query `Diagnostics` for given operator through pipeline?

Reply:
Again `Diagnostics` != first-class API to expose operators internals. It is fine for test and debugging, but let us not transform DALI classes into python dictionaries.
I can adjust the API to pass the name directly to the pipeline.

Reply:
Didn't notice that you responded, should have waited with pushing comments till I got a full picture.

The first case still stands, this is a lot of boilerplate, and everything should be made into single loop over all ops (if we want it to stay this way).

`Diagnostics` may not be first-class API, but we basically want such access to reader. I still don't think it's worth to put all of those functions in the Pipeline API and Operator API if they are related only to readers.

I suggest to either pack that information and do a bit more generic solution similar to `Diagnostics` or maybe register such markers with schema and make it accessible from Python. There are several solution, especially if we don't want to put 10 new functions to all the APIs for every possible reader flag.

Reply:
Done